ext/src/ruby_api/store.rs

```diff
@@ -363,3 +366,70 @@ pub fn init() -> Result<(), Error> {
 
     Ok(())
 }
+
+/// A resource limiter proxy used to report memory usage to Ruby's GC.
+struct TrackingResourceLimiter {
+    inner: StoreLimits,
+    allocated: isize,
+}
+
+impl TrackingResourceLimiter {
+    /// Create a new tracking limiter around an inner limiter.
+    pub fn new(resource_limiter: StoreLimits) -> Self {
+        Self {
+            inner: resource_limiter,
+            allocated: 0,
+        }
+    }
+}
+
+impl ResourceLimiter for TrackingResourceLimiter {
+    fn memory_growing(
+        &mut self,
+        current: usize,
+        desired: usize,
+        maximum: Option<usize>,
+    ) -> anyhow::Result<bool> {
+        let res = self.inner.memory_growing(current, desired, maximum);
+        if res.is_ok() && maximum.map_or(true, |maximum| desired <= maximum) {
+            self.allocated = desired as isize;
+            TrackingAllocator::adjust_memory_usage((desired - current) as isize);
+        }
+        res
+    }
+
+    fn table_growing(
+        &mut self,
+        current: u32,
+        desired: u32,
+        maximum: Option<u32>,
+    ) -> anyhow::Result<bool> {
+        self.inner.table_growing(current, desired, maximum)
+    }
+
+    fn memory_grow_failed(&mut self, error: anyhow::Error) -> anyhow::Result<()> {
+        self.inner.memory_grow_failed(error)
+    }
+
+    fn table_grow_failed(&mut self, error: anyhow::Error) -> anyhow::Result<()> {
+        self.inner.table_grow_failed(error)
+    }
+
+    fn instances(&self) -> usize {
+        self.inner.instances()
+    }
+
+    fn tables(&self) -> usize {
+        self.inner.tables()
+    }
+
+    fn memories(&self) -> usize {
+        self.inner.memories()
+    }
+}
+
+impl Drop for TrackingResourceLimiter {
```

Comment:
I believe you can rely on `ManuallyTracked`, which already takes care of doing this for you: https://docs.rs/rb-sys/latest/src/rb_sys/tracking_allocator.rs.html#195, similar to what was done in the original PR https://github.com/bytecodealliance/wasmtime-rb/pull/187/files#diff-b8e1c7a63c88d8176a489815defd575b01f6a596a617bb8561bb4fe54ac808d1R15

Reply:
Ah, I see `ManuallyTracked` has a `new` method (as opposed to `wrap` which requires an inner item). I can make that change.